src/rpucuda/cuda/rpucuda_JART_v1b_device.cu

```diff
@@ -0,0 +1,577 @@
+    T device_specific_Ndisc_max_bound_cuda = par_4.z;                          // [2]
+
+    /* NOTE: These values do random walks,
+             use refference to change the recorded values.
+    */ 
+    // T &device_specific_Ndiscmin_cuda = par_4.y; // [1]
+    // T &device_specific_Ndiscmax_cuda = par_4.w; // [3]
+    // T &device_specific_ldisc_cuda = par_2.x; // [0]
+    // T &device_specific_A_cuda = par_2.y; // [1]
+    T device_specific_Ndiscmin_cuda = par_4.y; // [1]
+    T device_specific_Ndiscmax_cuda = par_4.w; // [3]
+    T device_specific_ldisc_cuda = par_2.x; // [0]
+    T device_specific_A_cuda = par_2.y; // [1]
+
+    T &w = apparent_weight;
+    T &Ndisc = persistent_weight;
+
+    uint32_t pulse_counter = uint32_t (pulse_length/base_time_step);
+    // n is larger 0 in any case
+    pulse_counter = pulse_counter *n;
+    double Ndisc_double = Ndisc;
+    
+    // TODO: BUG: Use device variable bounds will result in PyTorch not receving the updated weights.
+    // T max_bound = MIN(device_specific_Ndisc_max_bound_cuda, device_specific_Ndiscmax_cuda);
+    // T min_bound = MAX(device_specific_Ndisc_min_bound_cuda, device_specific_Ndiscmin_cuda);
+    T max_bound = MIN(Ndisc_max_bound, device_specific_Ndiscmax_cuda);
+    T min_bound = MAX(Ndisc_min_bound, device_specific_Ndiscmin_cuda);
+
+    if (negative > 0) {
+      if (Ndisc_double < max_bound){
+        for (int i_updates = 0; i_updates < pulse_counter; i_updates++) {
+          T I_mem = -alpha_SET-beta_SET/(pow((1.0+pow((c_SET/Ndisc),d_SET)),f_SET));
+
+          // NOTE: V_disk = I_mem*(ldisc/(V_disk_coefficient*A*Ndisc))
+          // NOTE: Eion = V_disk/ldisc
+          T Eion = I_mem/(V_disk_coefficient*device_specific_A_cuda*Ndisc_double);
+
+          // NOTE: T gamma = gamma_coefficient*Eion
+          T gamma = gamma_coefficient*Eion;
+          
+          // NOTE: V - V_series = V_disk+V_plug+V_Schottky
+          T V_other_than_series = pulse_voltage_SET - (I_mem*(RseriesTiOx + R0 + V_series_coefficient*I_mem*I_mem));
+
+          T Treal = T0 + I_mem*V_other_than_series*Rth_negative_coefficient/device_specific_A_cuda;
+
+          // NOTE: dWamin = dWa_f = dWa*(sqrt(1.0-pow(gamma,2.0))-(gamma*M_PI)/2+gamma*asin(gamma)) = dWa_mean - dWa_difference
+          // NOTE: dWamax = dWa_r = dWa*(sqrt(1.0-pow(gamma,2.0))+(gamma*M_PI)/2+gamma*asin(gamma)) = dWa_mean + dWa_difference
+          T dWa_mean = dWa*(sqrt(1.0-pow(gamma,2.0))+gamma*asin(gamma));
+          T dWa_difference = dWa*((gamma*M_PI)/2.0);
+
+          T denominator = PHYSICAL_PARAMETER_kb_over_e*Treal;
+
+          T c_v0 = (Nplug+Ndisc_double)/2.0;
+          T F_limit = 1.0-pow((Ndisc_double/device_specific_Ndiscmax_cuda),10.0);
+          T dNdt = -(c_v0*a_ny0*F_limit*(exp(-(dWa_mean - dWa_difference)/denominator)-exp(-(dWa_mean + dWa_difference)/denominator)))/device_specific_ldisc_cuda;
+
+          Ndisc_double = Ndisc_double + dNdt*base_time_step;
```

Comment:
To improve performance, you should put all calculations that are repeated in the loop out side of the loop. In particular if the involve only constant numbers. There are all not necessary and could even be done once on CPU.  

Also note that "pow" and "asin" and "exp" are quite expensive. They could be replaced with the fast __expf etc. or x*x if pow(x, 2).